ryhope/src/storage/pgsql/storages.rs

```diff
@@ -1320,39 +1523,355 @@ where
     T::Node: Sync + Clone,
     V: PayloadInDb,
 {
-    delegate::delegate! {
-        to self.wrapped.lock().unwrap() {
-            async fn rollback_to(&mut self, epoch: Epoch) -> Result<()>;
-        }
-    }
-
-    fn remove(&mut self, k: T::Key) -> impl Future<Output = Result<()>> + Send {
+    async fn remove(&mut self, k: T::Key) -> Result<()> {
         trace!("[{self}] removing {k:?} from cache");
-        self.wrapped.lock().unwrap().nodes_cache.insert(k, None);
-        async { Ok(()) }
+        self.wrapped.write().await.nodes_cache.insert(k, None);
+        Ok(())
     }
 
-    fn update(&mut self, k: T::Key, new_value: V) -> impl Future<Output = Result<()>> + Send {
+    async fn update(&mut self, k: T::Key, new_value: V) -> Result<()> {
         trace!("[{self}] updating cache {k:?} -> {new_value:?}");
         // If the operation is already present from a read, replace it with the
         // new value.
         self.wrapped
-            .lock()
-            .unwrap()
+            .write()
+            .await
             .payload_cache
             .insert(k, Some(CachedValue::Written(new_value)));
-        async { Ok(()) }
+        Ok(())
     }
 
-    fn store(&mut self, k: T::Key, value: V) -> impl Future<Output = Result<()>> + Send {
+    async fn store(&mut self, k: T::Key, value: V) -> Result<()> {
         trace!("[{self}] storing {k:?} -> {value:?} in cache",);
         // If the operation is already present from a read, replace it with the
         // new value.
         self.wrapped
-            .lock()
-            .unwrap()
+            .write()
+            .await
             .payload_cache
             .insert(k, Some(CachedValue::Written(value)));
-        async { Ok(()) }
+        Ok(())
+    }
+
+    async fn rollback_to(&mut self, epoch: UserEpoch) -> Result<()> {
+        let inner_epoch = self
+            .wrapped
+            .read()
+            .await
+            .epoch_mapper
+            .to_incremental_epoch(epoch)
+            .await;
+        self.wrapped.write().await.rollback_to(inner_epoch).await
+    }
+
+    async fn rollback(&mut self) -> Result<()> {
+        let inner_epoch = self.wrapped.read().await.current_epoch();
+        ensure!(inner_epoch > 0, "cannot rollback past the initial epoch");
+        self.wrapped.write().await.rollback_to(inner_epoch).await
+    }
+}
+
+pub(crate) const INITIAL_INCREMENTAL_EPOCH: IncrementalEpoch = 0;
+
+#[derive(Clone, Debug)]
+pub struct EpochMapperStorage {
```

Comment:
Comments please.

Reply:
And can we put this whole thing into another file ? we're already at 1.5k lines.

Reply:
Done in commit https://github.com/Lagrange-Labs/mapreduce-plonky2/pull/425/commits/9da737de26934d27b84d971750a44f3a276ebce3